### PR TITLE
Upgrade jsoo-react

### PR DESCRIPTION
**Pull request by sim642**
This is the original version of #8.

---

Unfortunately jsoo-react has lots of breaking changes in the meanwhile. I think I managed to fix most of the compilation error, but there's one left which I don't really understand. There seems to be some change to the children representation, which requires weird `...` unpacking on all out custom components.

### TODO
- [x] Fix `ErrorBoundary` usage compilation.

### CODE REVIEW

**Comment by keremc** [comment]
If I remember correctly, I had the same issue with child components when I tried to update to the latest version of jsoo-react a few months ago. I think #8 should be fine for now. jsoo-react isn't stable yet anyway, so this probably won't be the last breaking change they make.

**Comment by sim642** [comment]
By the way, the also mentioned that there might be a compatibility library for this big breaking change at some point: https://github.com/ml-in-barcelona/jsoo-react/pull/162#issuecomment-1082301017.

**Comment by stilscher** [comment]
I rebased Simmo's work on a more recent version of the master branch, removed the opam pin such that we should now be using the opam release of jsoo-react directly, fixed the above mentioned compilation error with the `ErrorBoundary` and the remaining compilation errors. So while it at least builds, there is this weird `TypeError: c1 is undefined` error when running GobView and so far I have not really an idea why. Also, I think there might be a syntactically nicer way of fixing the ErrorBoundary compilation error, but so far I have not found out how.

**Comment by sim642** [comment]
> So while it at least builds, there is this weird `TypeError: c1 is undefined` error when running GobView and so far I have not really an idea why.

If you use the `dev` profile in `dune-workspace` to build, then js_of_ocaml should produce non-minimized JS that is somewhat more readable and has roughly original OCaml names for things.

**Comment by michael-schwarz** [comment]
@sim642 Since you worked on this initially, it might make sense if you review it. (I cannot request you as you are the one who started the PR.)

File: gobview.opam.locked [hunk 1]
```diff
@@ -46,13 +47,13 @@ depends: [
   "ctypes_stubs_js" {= "0.1"}
   "domain-name" {= "0.4.0"}
   "dune" {= "3.6.1"}
-  "dune-build-info" {= "3.6.1"}
+  "dune-build-info" {= "3.6.1" & dev}
```
Comment (michael-schwarz) [inline gobview.opam.locked:50]: Why are you adding these `dev` flags here? This seems a bit strange.

File: gobview.opam.locked [hunk 2]
```diff
@@ -89,19 +88,17 @@ depends: [
...
   "ocamlbuild" {= "0.14.2"}
   "ocamlfind" {= "1.9.5"}
-  "ocamlformat" {= "0.24.1" & dev}
-  "ocp-indent" {= "1.8.1"}
+  "ocamlformat" {= "0.25.1" & dev}
+  "ocamlformat-lib" {= "0.25.1" & dev}
+  "ocp-indent" {= "1.8.1" & dev}
   "ocplib-endian" {= "1.2"}
   "octavius" {= "1.2.2"}
-  "odoc" {= "2.2.0" & with-doc}
```
Comment (michael-schwarz) [inline gobview.opam.locked:99]: Why did you remove `odoc`? It still listed as a dependency in the `opam` file.

File: src/ui/base/select.re [hunk 3]
```diff
@@ -1,8 +1,9 @@
+open React.Dom.Dsl.Html;
 open Batteries;
 
 [@react.component]
 let make = (~options, ~compare=?, ~value, ~on_change) => {
-  let compare = Utils.fix_opt_arg(compare) |> Option.default((==));
+  let compare = Option.default((==), compare);
```
Comment (michael-schwarz) [inline src/ui/base/select.re:6]: Is comparing for reference equality the right thing to here ?

File: src/ui/content/gvFileView.re [hunk 4]
```diff
@@ -166,13 +166,13 @@ let make_editor = (goblint, warnings, file, contents, line, dispatch) => {
       ~render_validation_decorations="on",
       (),
     );
-  let state = React.useRef(None);
+  let state = React.use_ref(None);
   let on_mount = e =>
     {editor: e, decorations: [], content_widgets: [], view_zones: []}
     |> Option.some
```
Comment (michael-schwarz) [inline src/ui/content/gvFileView.re:172]: Is it maybe more idiomatic to

File: src/ui/content/gvFuncView.re [hunk 5]
```diff
@@ -20,21 +20,22 @@ let make = (~func: GvDisplay.func, ~dispatch) => {
 
   switch (func.dot) {
   | Some(dot) =>
+
     <ErrorBoundary
       message={
         "Cannot display the function graph. The generated DOT file is probably too large."
-        |> Js.string
-      }>
+        |> Js.string}>
```
Comment (michael-schwarz) [inline src/ui/content/gvFuncView.re:27]: In other places the other indentation is used.

File: src/ui/search/searchKindBuilder.re [hunk 6]
```diff
@@ -10,7 +11,7 @@ let options = [
 [@react.component]
 let make = (~value, ~on_change) => {
   <div className="mb-3">
-    <Label> {"Type" |> React.string} </Label>
+    <Label> ...{"Type" |> React.string} </Label>
```
Comment (michael-schwarz) [inline src/ui/search/searchKindBuilder.re:14]: Is there no more concise way to do this? Seems overkill for things that are constants?

File: gobview.opam.locked (hunk 1 above)
Comment (sim642) [reply gobview.opam.locked:50]: I'll retry to generate these files again such that they are consistent and also consistent with Goblint's. I think it's a bit tricky, so I probably should document it as well.

File: src/ui/search/searchKindBuilder.re (hunk 6 above)
Comment (stilscher) [reply src/ui/search/searchKindBuilder.re:14]: The call to `React.string` is required to convert the string to a React element. I am however not entirely sure why the spreading operator is required here. It is not used in the jsoo-react [Hello World example](https://ml-in-barcelona.github.io/jsoo-react/hello-world-reason) and React.string should return just one element, but when one removes it, a type error is thrown (expected React.element but found 'a list).

File: src/ui/base/select.re (hunk 3 above)
Comment (stilscher) [reply src/ui/base/select.re:6]: In Reason == and != is used for structural (in-)equality and === and !== is used for reference (in-)equality.

**Comment by sim642** [comment]
This seems to work fine with the `dev` profile, but with a non-`dev` profile (like `trace`), GobView breaks with an exception in the browser console:
```
Fatal error: exception Invalid_argument("Cilfacade.get_ikind: non-integer type void ")
```
Or maybe this profile issue exists already before this PR?

**Comment by stilscher** [comment]
I discovered this warning before but was not able reconstruct it. If I remember correctly, it was due to `get_ikind` (indirectly) being used for the definition of a value. And I suspect that it depended on the compilation order whether this caused an error or not.

**Comment by sim642** [comment]
I checked that for me it also happens on `master`, so I opened #38 about it, so this isn't a blocker for this PR, but still needs a fix.
